src/Data/Argonaut/Decode/Class.purs

```diff
@@ -66,6 +70,23 @@ instance decodeJsonString :: DecodeJson String where
 instance decodeJsonJson :: DecodeJson Json where
   decodeJson = Right
 
+
+toNonEmpty :: forall a f. (Plus f) => ({ head :: f a -> Maybe a, tail :: f a -> Maybe (f a) }) -> (f a) -> Either String (NonEmpty f a)
+toNonEmpty i a = case (Tuple (i.head a) (i.tail a)) of
```

Comment:
This could be cleaned up a little bit. Some of the parentheses are unnecessary, as is the tuple. Here's another way to write the case statement:

```purescript
toNonEmpty :: forall a f. Plus f => { head :: f a -> Maybe a, tail :: f a -> Maybe (f a) } -> f a -> Either String (NonEmpty f a)
toNonEmpty i a = case i.head a, i.tail a of
  Nothing, _ -> Left " is empty."
  Just h, Nothing -> Right $ singleton h
  Just h, Just t -> Right $ h :| t
```

As another note: `Data.Array` and `Data.List` already have functions that split apart the  structure into a head and tail called `uncons`:

https://pursuit.purescript.org/packages/purescript-lists/5.3.0/docs/Data.List#v:uncons
https://pursuit.purescript.org/packages/purescript-arrays/5.1.0/docs/Data.Array#v:uncons

For example:

```purescript
uncons :: forall a. List a -> Maybe { head :: a, tail :: List a }
```

Perhaps these could be re-used. I'm not sure this `toNonEmpty` function should be exported from the module. What do you think of decoding the array or list, then using `uncons` in each instance to create the non empty list, with a small amount of duplication of error messages in the two instances?